### PR TITLE
Refactor Command History

### PR DESCRIPTION
**Pull request by benedictkhoomw**
## What this does
* Moves `ViewHistoryCommand` and `ViewHistoryCommandParser` into their own sub-packages within `seedu.address.logic.commands` and `seedu.address.logic.parser` respectively. This will make the namespace neater.
* Refactors the command history selection logic out of the UI class `CommandBox` and into its own interface `CommandHistorySelector` backed by the class `SuppliedCommandHistorySelector`. The logic and state of command history selection is not user interaction, so it belongs in the Logic component, not the UI component.

### Command History Selection
1. Launch SunRez
2. Perform some normal commands.
3. Press the UP and DOWN arrow keys while the command box is in focus.
4. SunRez should navigate through your command history.
5. Hit ENTER at any point to run the currently selected command.

The history selection functionality should be unchanged from before the refactor.

## Notes
* This also modifies `Logic` and `LogicManager` to support a `getCommandHistorySelector()` which is used by `MainWindow` to initialize its `CommandBox` with the `CommandHistorySelector`.

### CODE REVIEW

**Review by stein414 — COMMENTED** [review]
Great work. The refactoring of the Command History to a dedicated history controller makes sense since the undo/redo feature can interface with the controller instead of being bundled together in the UI side.

**Review by stein414 — APPROVED** [review]
LGTM

**Comment by stein414** [comment]
I would like to discuss if the command history should also include failed commands. I think there is merit to mimic the same behaviour where the invalid command can be access via up and down arrow keys so that a Power User can edit the invalid command rather than retype it all.

If we do go down this path, I think a possible implementation is that each history entry will keep a field to keep the corresponding command used. Thus, if the field is null, it indicates an invalid command, otherwise it indicates a valid command.

File: src/main/java/seedu/address/logic/commandhistory/SuppliedCommandHistorySelector.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    public Optional<String> selectNext() {
+        return selectAt(commandHistoryIndex + 1);
+    }
+
+    @Override
+    public Optional<String> selectPrevious() {
+        return selectAt(commandHistoryIndex - 1);
+    }
+
+    private Optional<String> selectAt(int index) {
```
Comment (cnlinh) [inline src/main/java/seedu/address/logic/commandhistory/SuppliedCommandHistorySelector.java:45]: missing javadocs?

**Review by cnlinh — CHANGES_REQUESTED** [review]
Just missing 1 javadoc, otherwise LGTM

**Comment by benedictkhoomw** [comment]
> I would like to discuss if the command history should also include failed commands. I think there is merit to mimic the same behaviour where the invalid command can be access via up and down arrow keys so that a Power User can edit the invalid command rather than retype it all.
> 
> If we do go down this path, I think a possible implementation is that each history entry will keep a field to keep the corresponding command used. Thus, if the field is null, it indicates an invalid command, otherwise it indicates a valid command.

Hmm, thanks for the suggestion. SunRez currently does not consume the user's input when a command is invalid, so a power user can already edit an invalid command easily.

File: src/main/java/seedu/address/logic/commandhistory/SuppliedCommandHistorySelector.java (hunk 1 above)
Comment (benedictkhoomw) [reply src/main/java/seedu/address/logic/commandhistory/SuppliedCommandHistorySelector.java:45]: Good catch, thanks.

**Comment by stein414** [comment]
> > I would like to discuss if the command history should also include failed commands. I think there is merit to mimic the same behaviour where the invalid command can be access via up and down arrow keys so that a Power User can edit the invalid command rather than retype it all.
> > If we do go down this path, I think a possible implementation is that each history entry will keep a field to keep the corresponding command used. Thus, if the field is null, it indicates an invalid command, otherwise it indicates a valid command.
> 
> Hmm, thanks for the suggestion. SunRez currently does not consume the user's input when a command is invalid, so a power user can already edit an invalid command easily.

I see, that makes sense. In that case, the suggested feature seems to be not needed.

**Review by cnlinh — APPROVED** [review]

**Pull request merged**